Problem Set 6: confidence intervals and hypothesis testing in multiple regression<br>
Jyoti Chandrawanshi<br>

In [12]:
import numpy as np
import pandas as pd
import statsmodels.api as sm
import math
import scipy.stats as stats

import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
sns.set_style('white')

In [13]:
df = pd.read_csv("lead_mortality.csv")

In [14]:
# First we'll take a look at the data. Are there any null values that we have to worry about?
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 172 entries, 0 to 171
Data columns (total 15 columns):
year             172 non-null int64
city             172 non-null object
state            172 non-null object
age              172 non-null float64
hardness         172 non-null int64
ph               172 non-null float64
infrate          172 non-null float64
typhoid_rate     172 non-null float64
np_tub_rate      172 non-null float64
mom_rate         172 non-null float64
population       172 non-null float64
precipitation    172 non-null float64
temperature      172 non-null float64
lead             172 non-null int64
foreign_share    172 non-null float64
dtypes: float64(10), int64(3), object(2)
memory usage: 20.3+ KB


#### 1. What is the average infant mortality rate (Infrate) for cities with lead pipes and for cities with nonleaded pipes. Is there a statistically significant difference in the averages? 

In [19]:
y_group1 = df[df["lead"] == 1]['infrate']
y_group1_mean = y_group1.mean()

y_group2 = df[df["lead"] == 0]['infrate']
y_group2_mean = y_group2.mean()
    
 
mean_diff = y_group1_mean - y_group2_mean  


print("The Infrate for group 1(with lead pipes) is", round(y_group1_mean, 5), 
      ", the Infrate for group 2(with nonlead pipes) is", round(y_group2_mean, 5),
     "and the difference is", round(mean_diff, 5))

The Infrate for group 1(with lead pipes) is 0.40326 , the Infrate for group 2(with nonlead pipes) is 0.38117 and the difference is 0.02209


In [16]:
se_group1 = stats.sem(y_group1)
se_group2 = stats.sem(y_group2)

print("The standard error for group 1 is", round(se_group1, 5), 
      "and the standard error for group 2 is", round(se_group2, 5))

se_diff = math.sqrt(se_group1**2 + se_group2**2)
#Hint: use math.sqrt for square roots, and use "**" for exponents

print("The standard error of the difference is", round(se_diff, 5))

The standard error for group 1 is 0.01415 and the standard error for group 2 is 0.01992
The standard error of the difference is 0.02444


t = 0.02209 / 0.02444<br>
t = .90384615<br>
The average infant mortality rate (Infrate) for cities with lead pipes is() and for cities with nonleaded pipes is ().Since the the t value is less than 1.96 critical value so, we can not reject the null hypothesis or we can say that the there is not enough evidenace supporting diffrence in average infant mortality rate for cities with lead and with nonlead pipes.

#### 2. The amount of lead leached from lead pipes depends on the chemistry of the water running through the pipes. The more acidic the water (the lower its pH), the more lead is leeched.

#### a. Run a regression of Infrate on Lead, pH, and the interaction term Lead * pH 

In [17]:
df.lead = df.lead.astype(float)
df['lead_ph'] = df['lead']*df['ph']

X = df[['lead','ph','lead_ph']]
Y = df["infrate"]
X = sm.add_constant(X.values)
model = sm.OLS(Y, X).fit(cov_type = 'HC1')
model.summary(xname = ['const', 'lead','ph','lead_ph'])

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                infrate   R-squared:                       0.272
Model:                            OLS   Adj. R-squared:                  0.259
Method:                 Least Squares   F-statistic:                     20.97
Date:                Fri, 05 Mar 2021   Prob (F-statistic):           1.37e-11
Time:                        19:50:21   Log-Likelihood:                 108.52
No. Observations:                 172   AIC:                            -209.0
Df Residuals:                     168   BIC:                            -196.5
Df Model:                           3                                         
Covariance Type:                  HC1                                         
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.9189      0.150      6.106      0.000       0.624       1.214
lead           0.4618      0.208      2.224      0.026       0.055       0.869
ph            -0.0752      0.021     -3.588      0.000      -0.116      -0.034
lead_ph       -0.0569      0.028     -2.025      0.043      -0.112      -0.002
==============================================================================
Omnibus:                        4.916   Durbin-Watson:                   1.946
Prob(Omnibus):                  0.086   Jarque-Bera (JB):                4.987
Skew:                           0.411   Prob(JB):                       0.0826
Kurtosis:                       2.861   Cond. No.                         252.
==============================================================================

Warnings:
[1] Standard Errors are heteroscedasticity robust (HC1)
"""

#### b. The regression has four coefficients, including the intercept. Explain what each coefficient means. Does Lead have a statistically significant effect on infant mortality?

**Constant**: it's a intercept of infant mortality rate. If lead , ph and lead_ph are zeros the mean of infant mortility rate is 0.9189. <br>

**Coefficient of lead**: The mean change in lead by one unit will change infant mortality by 0.4618. when other variables are constant.also we can see that presence of lead will increase the infant mortality rate. p-value of lead is 0.026 which is less than .05 so we can say that lead have a statistically significant effect on infant mortality<br>

**Coefficient of lead**:The mean change in ph by one unit will decrease the infant mortality by 0.0752. when other variables are constant.also we can see that presence of ph will decrease the infant mortality rate.<br>

**Coefficient of lead_ph**:The mean change in ph * lead by one unit will decrease infant mortality by 0.0569 when other variables are constant. 


The mean change in ph by one unit will decrease infant mortality by 0.0752. when other variables are constant.The mean change in lead by one unit will change infant mortality by 0.4618. when other variables are constant. For mean change in ph * lead by one unit will decrease infant mortality by 0.0569 when other variables are constant. When ph and lead are not present or zero then the infant mortality is 0.9189. 

#### c. Does the effect of Lead on infant mortality depend on pH? Is this dependence statistically significant? 

From regression equation: <br>

Infant mortality = 0.9189 + 0.4618 * lead - 0.0752	* ph - 0.0569 * lead_ph<br>

Yes lead_ph has statistically significant effect on infant mortality because the p-value of interaction term lead_ph (0.043) is less than alpha at 0.05.

#### 3. The analysis in question 2 may suffer from omitted variable bias because it doesn't include other factors that affect infant mortality that may also be correlated with Lead and pH. When looking at other variables in the dataset, explain why you think the regression does or does not have omitted variable bias. 

In [18]:
X = df[['lead','ph','age','hardness','typhoid_rate','np_tub_rate','mom_rate','population','precipitation','temperature','foreign_share']]
Y = df["infrate"]
X = sm.add_constant(X.values)
model = sm.OLS(Y, X).fit(cov_type = 'HC1')
model.summary(xname = ['const', 'lead','ph','age','hardness','typhoid_rate','np_tub_rate','mom_rate','population','precipitation','temperature','foreign_share'])

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                infrate   R-squared:                       0.549
Model:                            OLS   Adj. R-squared:                  0.518
Method:                 Least Squares   F-statistic:                     16.42
Date:                Fri, 05 Mar 2021   Prob (F-statistic):           2.75e-21
Time:                        19:50:21   Log-Likelihood:                 149.79
No. Observations:                 172   AIC:                            -275.6
Df Residuals:                     160   BIC:                            -237.8
Df Model:                          11                                         
Covariance Type:                  HC1                                         
=================================================================================
                    coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------
const             0.8548      0.243      3.524      0.000       0.379       1.330
lead              0.0318      0.019      1.648      0.099      -0.006       0.070
ph               -0.0613      0.015     -4.202      0.000      -0.090      -0.033
age              -0.0207      0.005     -4.401      0.000      -0.030      -0.011
hardness         -0.0001   8.11e-05     -1.312      0.190      -0.000    5.25e-05
typhoid_rate      0.9907      0.326      3.040      0.002       0.352       1.630
np_tub_rate       1.8297      0.752      2.435      0.015       0.357       3.303
mom_rate          0.0055      0.304      0.018      0.985      -0.591       0.602
population    -1.538e-06   1.89e-06     -0.812      0.417   -5.25e-06    2.17e-06
precipitation     0.0708      0.017      4.158      0.000       0.037       0.104
temperature       0.0037      0.002      2.075      0.038       0.000       0.007
foreign_share     0.2583      0.102      2.540      0.011       0.059       0.458
==============================================================================
Omnibus:                        2.845   Durbin-Watson:                   1.796
Prob(Omnibus):                  0.241   Jarque-Bera (JB):                2.473
Skew:                           0.197   Prob(JB):                        0.290
Kurtosis:                       3.435   Cond. No.                     2.90e+05
==============================================================================

Warnings:
[1] Standard Errors are heteroscedasticity robust (HC1)
[2] The condition number is large, 2.9e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""